### PR TITLE
Make the order of symbols returned from DataFlowAnalysis deterministic

### LINKED ISSUE #803: Order of symbols provided by DataFlowAnalysis should be deterministic
DataFlowAnalysis exposes various symbol lists (VariablesDeclared, DataFlowsIn, etc.). The order of symbols in these arrays is non-deterministic. The underlying implementation usually dumps the content of a hash set or a dictionary to these arrays.

Since it’s too easy to take a dependency on the order when enumerating items of an array.

### PR DESCRIPTION
**Pull request by gafter**
and adjust tests for deterministic region analysis
Fixes #803

### CODE REVIEW

**Comment by gafter** [comment]
@jaredpar @VSadov @AlekseyTs @agocke @VladimirReshetnikov Could a couple of you review this please?

File: src/Compilers/CSharp/Portable/Symbols/Symbol.cs [hunk 1]
```diff
@@ -240,18 +240,21 @@ public bool IsDefinition
...
         /// if nothing is cached yet.)
         /// </para>
         /// <para>
-        /// Only members of source namespaces, original source types, and namespaces that can be merged
-        /// need implement this function.
+        /// Only (original) source symbols and namespaces that can be merged
+        /// need implement this function if they want to do so for efficiency.
         /// </para>
         /// </summary>
         internal virtual LexicalSortKey GetLexicalSortKey()
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Symbols/Symbol.cs]: This used to be used for type symbols and namespaces only. Now it is also used for locals, range variables, and parameters. This implementation is good for most symbol kinds.

**Comment by gafter** [comment]
@jaredpar @VSadov @AlekseyTs @agocke @VladimirReshetnikov Could a couple of you review this please?

**Comment by gafter** [comment]
@mikedn All excellent comments, thank you! I will submit a revision of the code review that addresses them.

**Comment by gafter** [comment]
@mikedn I believe your comments have been directly addressed in this revision of the pull request. Any further comments you have are most welcome!

@jaredpar @VSadov @AlekseyTs @agocke @VladimirReshetnikov Could a couple of you review this please?

**Comment by mikedn** [comment]
`Sort` still boxes the immutable arrays from `AnalyzeReadWrite`. It doesn't look like this code is performance critical as it's only used by the extract method refactoring so I wouldn't change this.

If you think that performance is a valid concern here then you could add a `Sort(ImmutableArray<Symbol> array)` overload which simply returns `array.Sort(LexicalOrderSymbolComparer.Instance)`.

Either way, the casts are gone so LGTM, thank you!

**Comment by gafter** [comment]
@mikedn I don't understand your comment about `Sort`. Your proposed change would still box the immutable arrays from `AnalyzeReadWrite`, wouldn't it? Since the arrays are immutable, sorting them requires copying them.

**Comment by mikedn** [comment]
@gafter `ImmutableArray<T>` is a struct so it will get boxed when passed to `Sort(IEnumerable<Symbol>)`. And yes, sorting requires allocating another array but that's not related to struct boxing.

Speaking of copying, in this particular case 2 copies are made: one from the original array to the builder array and another one from the builder array to the new array. If you sort the immutable array directly then only one copy will be made, from the original array to the new array.

**Comment by gafter** [comment]
@mikedn I understand now. I think this latest revision is better. What do you think?

**Comment by gafter** [comment]
@jaredpar @VSadov @AlekseyTs @agocke @VladimirReshetnikov Could a couple of you review this please?

**Comment by mikedn** [comment]
@gafter I think that it's even better than my Sort overload. :+1:

File: src/Compilers/CSharp/Portable/CSharpCodeAnalysis.csproj [hunk 2]
```diff
@@ -348,7 +348,7 @@
     <Compile Include="FlowAnalysis\AlwaysAssignedWalker.cs" />
     <Compile Include="FlowAnalysis\ControlFlowAnalysis.cs" />
     <Compile Include="FlowAnalysis\ControlFlowPass.cs" />
-    <Compile Include="FlowAnalysis\DataFlowAnalysis.cs" />
+    <Compile Include="FlowAnalysis\CSharpDataFlowAnalysis.cs" />
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/CSharpCodeAnalysis.csproj]: Is there a particular reason for the rename? It feels strange to use CSharp prefix for files in CSharp folder.

File: src/Compilers/CSharp/Portable/FlowAnalysis/ReadWriteWalker.cs [hunk 3]
```diff
@@ -28,17 +28,17 @@ internal class ReadWriteWalker : AbstractRegionDataFlowPass
                 walker.Analyze(ref badRegion);
                 if (badRegion)
                 {
-                    readInside = writtenInside = readOutside = writtenOutside = captured = unsafeAddressTaken = ImmutableArray<ISymbol>.Empty;
+                    readInside = writtenInside = readOutside = writtenOutside = captured = unsafeAddressTaken = ImmutableArray<Symbol>.Empty;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/ReadWriteWalker.cs]: Enumerable.Empty<Symbol>()?

File: src/Compilers/CSharp/Portable/FlowAnalysis/ReadWriteWalker.cs [hunk 4]
```diff
@@ -28,17 +28,17 @@ internal class ReadWriteWalker : AbstractRegionDataFlowPass
...
-                    writtenOutside = ((IEnumerable<ISymbol>)walker._writtenOutside).ToImmutableArray();
+                    readInside = walker._readInside;
+                    writtenInside = walker._writtenInside;
+                    readOutside = walker._readOutside;
+                    writtenOutside = walker._writtenOutside;
 
-                    captured = ((IEnumerable<ISymbol>)walker.GetCaptured()).ToImmutableArray();
-                    unsafeAddressTaken = ((IEnumerable<ISymbol>)walker.GetUnsafeAddressTaken()).ToImmutableArray();
+                    captured = walker.GetCaptured();
+                    unsafeAddressTaken = walker.GetUnsafeAddressTaken();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/ReadWriteWalker.cs]: Should we ensure that the HashSets cannot be modified?

File: src/Compilers/CSharp/Portable/Symbols/Symbol.cs [hunk 5]
```diff
@@ -240,18 +240,21 @@ public bool IsDefinition
...
         {
-            throw ExceptionUtilities.Unreachable;
+            var locations = this.Locations;
+            var declaringCompilation = this.DeclaringCompilation;
+            Debug.Assert(declaringCompilation != null); // require that it is a source symbol
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Symbol.cs]: Consider changing it to a release check that throws.

File: src/Compilers/VisualBasic/Portable/BasicCodeAnalysis.vbproj [hunk 6]
```diff
@@ -101,7 +101,7 @@
     <Compile Include="Analysis\FlowAnalysis\ControlFlowAnalysis.vb" />
     <Compile Include="Analysis\FlowAnalysis\ControlFlowPass.LocalState.vb" />
     <Compile Include="Analysis\FlowAnalysis\ControlFlowPass.vb" />
-    <Compile Include="Analysis\FlowAnalysis\DataFlowAnalysis.vb" />
+    <Compile Include="Analysis\FlowAnalysis\VisualBasicDataFlowAnalysis.vb" />
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/BasicCodeAnalysis.vbproj]: Same question about rename.

File: src/Compilers/CSharp/Portable/CSharpCodeAnalysis.csproj (hunk 2 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/CSharpCodeAnalysis.csproj]: Yes; I changed the name of the file to reflect the name of the class it contains.

File: src/Compilers/CSharp/Portable/FlowAnalysis/ReadWriteWalker.cs (hunk 4 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/FlowAnalysis/ReadWriteWalker.cs]: As written that is the responsibility of the caller. That is a contract between this method and its only client.

File: src/Compilers/VisualBasic/Portable/BasicCodeAnalysis.vbproj (hunk 6 above)
Comment (gafter) [inline src/Compilers/VisualBasic/Portable/BasicCodeAnalysis.vbproj]: Changed the file name to correspond to the class within it.

File: src/Compilers/CSharp/Portable/FlowAnalysis/ReadWriteWalker.cs (hunk 3 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/FlowAnalysis/ReadWriteWalker.cs]: Yes; it should be.

File: src/Compilers/CSharp/Portable/CSharpCodeAnalysis.csproj (hunk 2 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/CSharpCodeAnalysis.csproj]: I see. Thanks.

File: src/Compilers/CSharp/Portable/FlowAnalysis/ReadWriteWalker.cs (hunk 3 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/ReadWriteWalker.cs]: Resolved

File: src/Compilers/CSharp/Portable/FlowAnalysis/ReadWriteWalker.cs (hunk 4 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/ReadWriteWalker.cs]: I see. Thanks.

File: src/Compilers/CSharp/Portable/Symbols/Symbol.cs (hunk 5 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Symbol.cs]: Could you respond to this comment about the Assert please?

File: src/Compilers/VisualBasic/Portable/BasicCodeAnalysis.vbproj (hunk 6 above)
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/BasicCodeAnalysis.vbproj]: I see. Thanks.

File: src/Compilers/CSharp/Portable/Symbols/Symbol.cs (hunk 5 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Symbols/Symbol.cs]: As this is an internal contract, we generally prefer debug assertions.

**Comment by AlekseyTs** [comment]
LGTM

**Comment by gafter** [comment]
@jaredpar @VSadov @agocke @VladimirReshetnikov Could one (more) of you review this please?

**Comment by gafter** [comment]
@jaredpar @VSadov @agocke @VladimirReshetnikov Could one (more) of you review this please?

**Comment by agocke** [comment]
LGTM

File: src/Compilers/CSharp/Portable/FlowAnalysis/CSharpDataFlowAnalysis.cs [hunk 7]
```diff
@@ -292,5 +294,13 @@ public sealed override bool Succeeded
                 return _succeeded.Value;
             }
         }
+
+        internal static ImmutableArray<ISymbol> Sort(IEnumerable<Symbol> data)
+        {
+            var builder = ArrayBuilder<Symbol>.GetInstance();
+            builder.AddRange(data);
+            builder.Sort(LexicalOrderSymbolComparer.Instance);
+            return builder.ToImmutableAndFree().As<ISymbol>();
```
Comment (VladimirReshetnikov) [inline src/Compilers/CSharp/Portable/FlowAnalysis/CSharpDataFlowAnalysis.cs]: Is it possible that `data` contains 2 local variables with the same name declared in disjoint blocks?

Comment (gafter) [inline src/Compilers/CSharp/Portable/FlowAnalysis/CSharpDataFlowAnalysis.cs]: Indeed it is possible and the tests have a few examples of that.

**Pull request merged**